jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/TreeDynamicConnectivity.java

```diff
@@ -0,0 +1,629 @@
+/*
+ * (C) Copyright 2020-2020, by Timofey Chudakov and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * See the CONTRIBUTORS.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.connectivity;
+
+import org.jgrapht.util.AVLTree;
+import org.jgrapht.util.DoublyLinkedList;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.jgrapht.util.DoublyLinkedList.ListNode;
+import static org.jgrapht.util.AVLTree.TreeNode;
+
+/**
+ * Data structure for storing dynamic trees and querying node connectivity
```

Comment:
I assume you are using the Euler tour data structure, so we should point to the proper reference which is: 

```
Monika Rauch Henzinger, Valerie King: Randomized dynamic graph algorithms with polylogarithmic time per operation. STOC 1995: 519-527
```